File: Go/ExampleTests/common/framevalues_string.go
package common

type FrameValuesString struct {
	Frame string
	Roll1 string
	Roll2 string
	Roll3 string
	Score string
	TotalScore string
}

func NewFrameValuesStringFromSlice(v []string) FrameValuesString {
	s := FrameValuesString{}
	if len(v) > 0 { s.Frame = v[0] }
	if len(v) > 1 { s.Roll1 = v[1] }
	if len(v) > 2 { s.Roll2 = v[2] }
	if len(v) > 3 { s.Roll3 = v[3] }
	if len(v) > 4 { s.Score = v[4] }
	if len(v) > 5 { s.TotalScore = v[5] }
	return s
}

// NewFrameValuesStringFromText builds from the text form, e.g. Money as "25 USD".
func NewFrameValuesStringFromText(text string) FrameValuesString {
	parts := RequireTokens(text, 6, "FrameValues")
	return FrameValuesString{
		Frame: parts[0],
		Roll1: parts[1],
		Roll2: parts[2],
		Roll3: parts[3],
		Score: parts[4],
		TotalScore: parts[5],
	}
}

func (s FrameValuesString) String() string {
	return Token(s.Frame) + " " + Token(s.Roll1) + " " + Token(s.Roll2) + " " + Token(s.Roll3) + " " + Token(s.Score) + " " + Token(s.TotalScore)
}

func (s FrameValuesString) Equals(o FrameValuesString) bool {
	return DNCEqual(s.Frame, o.Frame) &&
		DNCEqual(s.Roll1, o.Roll1) &&
		DNCEqual(s.Roll2, o.Roll2) &&
		DNCEqual(s.Roll3, o.Roll3) &&
		DNCEqual(s.Score, o.Score) &&
		DNCEqual(s.TotalScore, o.TotalScore)
}

func EqualFrameValuesStringSlices(a, b []FrameValuesString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/framevalues_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct FrameValuesString {
    pub frame: String,
    pub roll1: String,
    pub roll2: String,
    pub roll3: String,
    pub score: String,
    pub totalscore: String,
}

impl FrameValuesString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            frame: v.get(0).copied().unwrap_or("").to_string(),
            roll1: v.get(1).copied().unwrap_or("").to_string(),
            roll2: v.get(2).copied().unwrap_or("").to_string(),
            roll3: v.get(3).copied().unwrap_or("").to_string(),
            score: v.get(4).copied().unwrap_or("").to_string(),
            totalscore: v.get(5).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 6, "FrameValues");
        Self {
            frame: parts[0].clone(),
            roll1: parts[1].clone(),
            roll2: parts[2].clone(),
            roll3: parts[3].clone(),
            score: parts[4].clone(),
            totalscore: parts[5].clone(),
        }
    }
}

impl std::fmt::Display for FrameValuesString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.frame), crate::common::tokens::token(&self.roll1), crate::common::tokens::token(&self.roll2), crate::common::tokens::token(&self.roll3), crate::common::tokens::token(&self.score), crate::common::tokens::token(&self.totalscore)].join(" "))
    }
}

impl PartialEq for FrameValuesString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.frame, &other.frame)
            && crate::common::dnc_equal(&self.roll1, &other.roll1)
            && crate::common::dnc_equal(&self.roll2, &other.roll2)
            && crate::common::dnc_equal(&self.roll3, &other.roll3)
            && crate::common::dnc_equal(&self.score, &other.score)
            && crate::common::dnc_equal(&self.totalscore, &other.totalscore)
    }
}
